### ISSUE TITLE
new endpoint for package metadata

### ISSUE THREAD

**Issue by GeorgSchneider** [digital-asset/daml#3508]
I've come across several use cases where I need information about the DAML-LF archive in my client-side code. Examples include:
- Display of available choices on a contract in the UI
- Generic parameter form for creating contracts or exercising choices

DABL currently maintains its own endpoint for this.

Information that should be provided:
- Types
- Templates
- Choices (+ parameters, non- / pre- / post-consuming)

I'd suggest to provide:
/packages: serves the ids of packages currently loaded
/package/{id}: serves the metadata for a given package

**Comment by leo-da** [comment]
cc @gerolf-da @bame-da

**Comment by leo-da** [comment]
cc @remyhaemmerle-da re: Interface Reader service

**Comment by bame-da** [comment]
I think it's useful for advanced use-cases, but not very high priority in my view. Does the DABL endpoint do everything you want @georg-da ? If so, maybe it could be upstreamed.

**Comment by garyverhaegen-da** [comment]
I have personally found [GraphiQL](https://github.com/graphql/graphiql) super valuable, even in a team of 3 controlling the entire codebase end-to-end. This is much more important for clients written in dynamic languages, as I suppose much of that information could otherwise be included in the type definitions from codegen.

**Comment by remyhaemmerle-da** [comment]
I do not think this is relevant only for the json api. It is a feature we will need in the ledger api itself.

**Comment by S11001001** [comment]
I do not think we should have this feature in the json-api.

The intent of the json-api is not to support all use cases, or even _many_ use cases, but rather to support basic _now_ active contract set functionality on _known_ contract types. (I admit this is my own characterization, but one I believe accurate based upon all design discussions so far.) That excludes metaprogramming of the type you describe; see [the doc intro](https://docs.daml.com/json-api/index.html).

In other words, if the json-api provides enough functionality to write navigator, then it has already failed at being a simple on-ramp to getting started writing applications against the ledger.

On the other hand, the feature is completely appropriate for the ledger API, against which of course all kinds of metaprogramming are use cases of interest.

**Comment by GeorgSchneider** [comment]
@S11001001 I agree that we definitely need package metadata on the Ledger API. But I also think we can easily provide a pass-through in the JSON API, same as we did for the party mgmt endpoints. If we can do party mgmt we also should be able to do pkg mgmt.

This doesn't hinder its goal of being an easy-to-use on-ramp, it simply makes it more powerful. The problem with having this only in gRPC is that I can't use it from a web UI, which is a use case I've encountered already several times.

**Comment by S11001001** [comment]
> But I also think we can easily provide a pass-through in the JSON API, same as we did for the party mgmt endpoints. If we can do party mgmt we also should be able to do pkg mgmt.

Parties were simple, but that does not mean packages are too. Parties are just lists of strings, and was easy to tack on to the API in #2982. By contrast, the metadata discussed here involves types, which entails a JSON representation of types, i.e. the daml-lf/interface API. There is nothing trivial or "pass-through" about this in the way that parties were.

**Comment by hurryabit** [comment]
We're not going to do this in the foreseeable future. This is not necessary for the recommended path. We'll cover a _static_ version of what this endpoint would offer by means of compile-time code generation.
There's also no strict need to integrate this with the JSON API. It could also be a standalone service implemented by others.
cc @bame-da @da-tanabe @georg-da
